### ISSUE TITLE
ufunc 'equal' doesn't return anything else than booleans

### ISSUE THREAD

**Issue by Celelibi** [numpy/numpy#10948]
I know it's pretty uncommon to have a class that implement a `__eq__` method that returns anything else than a boolean. But my use-case is that I build symbolic formulas.

Therefore, if I let numpy broadcast a `==` operator on an array of objects, it always return an array of bool, instead of an array of objects as I would have expected. Here is an example code that demonstrate this behavior.

```python
import numpy as np

class Comparable(object):
    def __eq__(self, other):
        return self

a = np.array([Comparable()])
print(a == 0)
```
Which prints:
```
[ True]
```
instead of:
```
[<__main__.Comparable object at 0x7f8f7c2c04e0>]
```

I found a workaround by subclassing `numpy.ndarray` and implementing `__array_ufunc__` to special-case the ufunc `equal`. But it's quite a pain to implement, and some details of the behavior remain unclear. Like, why the integer constant `0` is an instance of `np.int64` when `__array_ufunc__` is called, but when the ufunc calls `__eq__`, it's a python `int`.

If making `np.equal` return an array of objects would be too much of a backward incompatibility, maybe a solution could be found make the implementation of such workaround easier.

**Comment by eric-wieser** [comment]
This request came up recently. In 1.15, you can do `np.equal(a, 0, dtype=object)`

**Comment by Celelibi** [comment]
Thanks. Then I guess this issue will be closed when the version 1.15 is out.

**Comment by mhvk** [comment]
I think I'll close this issue now rather than let it hang around. @Celelibi - if you can test your code against master just to be sure it solves your problems, that would be helpful!
